src/Juvix/Core/Common/Context.hs

```diff
@@ -48,80 +63,371 @@ data Definition term ty sumRep
+--------------------------------------------------------------------------------
+type NameSymbol = NameSymbol.T
 
-(!?) :: T term ty sumRep -> Symbol -> Maybe (Definition term ty sumRep)
-(!?) = flip lookup
+nameSymbolToSymbol :: NameSymbol.T -> Symbol
+nameSymbolToSymbol = NameSymbol.toSymbol
+
+nameSymbolFromSymbol :: Symbol -> NameSymbol.T
+nameSymbolFromSymbol = NameSymbol.fromSymbol
+
+--------------------------------------------------------------------------------
+-- Body
+--------------------------------------------------------------------------------
 
+empty :: NameSymbol.T -> T term ty sumRep
+empty sym =
+  case addPathWithValue sym CurrentNameSpace fullyEmpty of
+    Lib.Left _ -> error "impossible"
+    Lib.Right x -> x
+  where
+    fullyEmpty =
+      ( T
+          { currentNameSpace = NameSpace.empty,
+            currentName = sym,
+            topLevelMap = HashMap.empty
+          }
+      )
+
+qualifyName :: NameSymbol.T -> T term ty sumRep -> NameSymbol.T
+qualifyName sym T {currentName} = currentName <> sym
+
+--------------------------------------------------------------------------------
+-- Functions on the Current NameSpace
+--------------------------------------------------------------------------------
+
+lookupCurrent ::
+  NameSymbol.T -> T term ty sumRep -> Maybe (NameSpace.From (Definition term ty sumRep))
+lookupCurrent =
+  lookupGen (\_ currentLookup -> currentLookup)
+
+-- TODO ∷ Maybe change
+-- By default add adds it to the public map by default!
 add ::
-  Symbol ->
+  NameSpace.From Symbol ->
   Definition term ty sumRep ->
   T term ty sumRep ->
   T term ty sumRep
-add sy term (T map) = T $ HashMap.insert sy term map
+add sy term t =
+  t {currentNameSpace = NameSpace.insert sy term (currentNameSpace t)}
 
 remove ::
-  Symbol -> T term ty sumRep -> T term ty sumRep
-remove sy (T map) = T $ HashMap.delete sy map
+  NameSpace.From Symbol -> T term ty sumRep -> T term ty sumRep
+remove sy t = t {currentNameSpace = NameSpace.remove sy (currentNameSpace t)}
+
+publicNames :: T term ty sumRep -> [Symbol]
+publicNames T {currentNameSpace} =
+  let NameSpace.List {publicL} = NameSpace.toList currentNameSpace
+   in fst <$> publicL
+
+toList :: T term ty sumRep -> NameSpace.List (Definition term ty sumRep)
+toList T {currentNameSpace} = NameSpace.toList currentNameSpace
+
+topList :: T term ty sumRep -> [(Symbol, Definition term ty sumRep)]
+topList T {topLevelMap} = HashMap.toList topLevelMap
 
-modify,
-  update ::
-    (Definition term ty sumRep -> Maybe (Definition term ty sumRep)) ->
-    Symbol ->
-    T term ty sumRep ->
-    T term ty sumRep
-modify f k (T map) = T (HashMap.update f k map)
-update = modify
+--------------------------------------------------------------------------------
+-- Global Functions
+--------------------------------------------------------------------------------
+-- we lose some type information here... we should probably reserve it somehow
```

Comment:
to be honest I find this architecture a little confusing - do we really need to transform the context every time we switch in and out of namespaces? can't we just have a pass which resolves everything to global (fully qualified) identifiers?

Reply:
or is this intended to be more efficient somehow?

Reply:
> to be honest I find this architecture a little confusing - do we really need to transform the context every time we switch in and out of namespaces? can't we just have a pass which resolves everything to global (fully qualified) identifiers?

We don't transform the context at all, we simply, move one out of the nested map, and onto the front, and move the one from the front back into the map itself.

> or is this intended to be more efficient somehow?

Yes this should be much faster, as instead of always adding a variables 5 levels deep (and thus reconstructing 5 levels of tree for every single change, instead we only change a single map in the front, and are able to have proper private semantics (if we had this global qualified version, how do I know when I'm legally allowed to touch the private variables?

Reply:
> this function needs a more detailed comment, when are we switching namespaces? what should the pre/post-conditions be?

We switch them every time we need to be in a new module/record. The Pre/Post is simply that the `CurrentNameSpace` tag moves, and the structure is properly reflected. If an invalid path is given `Foo.x.Y` where `Foo.x` is 3, then the function fails saying the operation has been aborted. No real complex pre-post conditions are here

Reply:
> Hmm OK, do we no longer have the ability to use . syntax with records? Can we add a record test?

It seems I can't comment on the comment above this, so I'll handle it here.

we can, records are namespaces after all, we'll need better detection, one issues we'll have (we have this in the current backend as is!), is that I can't tell what is a namespace or not without type checking. Pretty much all inner modules will be seen as definitions and fail somewhat. This is something we can't resolve too well. For all the contextify passes we don't really jump ahead and resolve things, so this will only error on module opens if it can't detect that we see a record.

Reply:
Further, this concept of the current namespace will be useful for the repl later, since it'll need to assume what namespace it's in, along with changing on the current one

Reply:
Another benefit is reaching below the current map, if we instead only had the single `top level` map and searching down to see the current map, then I'll need to do 5 derefs of work before I can even determine if what I'm searching for is below the current namespace or I should research at the top, making detection of local vs global much much more expensive

Reply:
> Yes this should be much faster, as instead of always adding a variables 5 levels deep (and thus reconstructing 5 levels of tree for every single change, instead we only change a single map in the front, and are able to have proper private semantics (if we had this global qualified version, how do I know when I'm legally allowed to touch the private variables?

Sure, that makes sense.

> Further, this concept of the current namespace will be useful for the repl later, since it'll need to assume what namespace it's in, along with changing on the current one

Yes, I agree with that.

Reply:
> We switch them every time we need to be in a new module/record. The Pre/Post is simply that the CurrentNameSpace tag moves, and the structure is properly reflected. If an invalid path is given Foo.x.Y where Foo.x is 3, then the function fails saying the operation has been aborted. No real complex pre-post conditions are here

OK, let's just add this as a comment. It is clear after spending several minutes reading the function, but not immediately, since the datatype names are not super indicative.

Reply:
Sure will do